### PR TITLE
[SPARK-19372][SQL] Fix throwing a Java exception at df.fliter() due to 64KB bytecode size limit

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

When an expression for `df.filter()` has many nodes (e.g. 400), the size of Java bytecode for the generated Java code is more than 64KB. It produces an Java exception. As a result, the execution fails.
This PR continues to execute by calling `Expression.eval()` disabling code generation if an exception has been caught.

## How was this patch tested?

Add a test suite into `DataFrameSuite`

### CODE REVIEW

**Comment by kiszk** [comment]
@davies, could you please review this?

**Comment by srowen** [comment]
Is it right to just do this in one code path? why not all similar cases where the 64k limit is exceeded?
Can the underlying problem be fixed or improved?
You're also catching all Exceptions, why not something more specific? is it possible?
It feels like a funny change to make for these reasons.

**Comment by kiszk** [comment]
I already identified where 64k limit is exceeded. As a result, I think that it is not easy to fix the issue. 
The previous issues are derived from independent blocks (in `Seq[ExprCode]`) that can be splited into multiple methods. On the other hand, it is not easy split a tree with many nodes (in `ExprCode`) that cannot be into multiple methods. It is very hard to split a `ExprCode` into multiple correct `ExprCode`s.

For catching exception, you are right. I have to catch only an exception related to compilation errors. I will address it by handling nested exceptions. I will address this.

**Comment by marmbrus** [comment]
I agree with the general approach of having a fallback from code generation to interpreted evaluation, but I also agree that this feels too narrowly targeted.  In particular, why do this in one operator rather than in `newPredicate` (or maybe even in codegen itself).

Another thing that maybe @davies can comment on, I thought we already had this fallback implemented? So I'm curious why its not already handling this test case.  Maybe there is an existing mechanism we just need to make more general.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -213,12 +217,30 @@ case class FilterExec(condition: Expression, child: SparkPlan)
...
+        val predicate = newPredicate(condition, child.output)
+        predicate.initialize(0)
+        iter.filter { row =>
+          val r = predicate.eval(row)
+          if (r) numOutputRows += 1
+          r
+        }
+      } catch {
+        // JaninoRuntimeException is in a nested exception if Java compilation error occurs
+        case e: Exception if ExceptionUtils.getRootCause(e).isInstanceOf[JaninoRuntimeException] =>
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Rather than walk the exception tree, should we just make [our wrapping](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala#L957) more specific?

Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Good catch. I will make this wrapping more specific.

**Comment by kiszk** [comment]
@marmbrus thank you for your comments.

For feedback mechanism, I imagine that you are talking about [this](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala#L357-L364). When the whole-stage codegen is enabled, this fallback works.
In this case, the whole-stage codegen is disabled since the number of fields are big and [these two variables](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala#L463-L466) will have `false`.

I agree that we should prepare more generic approach that can be applicable to more cases. However, it is not easy to implement for now. It would require some refactoring.
As one of simple refactoring, if an exception occurs, `newPredicate` returns `null` or `none`. Then, caller of `newPredicate` always checks return value for handling of code generation failure. It make caller code simpler.
While more complex refactoring could be applicable, another PR should address it.

What do you think?


FYI: I have just noticed that this problem may occur at [CartesianProductExec](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/CartesianProductExec.scala#L104-L110) and [InMemoryTableScanExec](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala#L136). I have to do the same thing in these two places.

**Comment by marmbrus** [comment]
I don't think we need a complex refactoring.  Why can't `newPredicate` catch the exception, log a warning and return an interpreted `Predicate`?

**Comment by kiszk** [comment]
I am refactoring `newPredicate`. `newPredicate` will catch the exception and log a warning. However, I think that `newPredicate` cannot return an interpreted result. This is because code generation return `Predicate` while ` BindReferences.bindReference(condition, child.output)` returns `Expression`. These two classes exists in different class hierarchy.

**Comment by kiszk** [comment]
@marmbrus I have just commit the code of intermediate refactoring. Would it be possible to give comments?
If it is fine (return `null` and check it at caller), I will update other caller sites for `newPredicate`.

**Comment by marmbrus** [comment]
There appears to have been some code drift (as `GeneratePredicate` and `InterpretedPredicate` both used to return a class that inherited from a common interface), but I don't think its hard to just isolate the fault handling in `newPredicate`.

Just fix [InterpretedPredicate](https://github.com/apache/spark/blob/9b7a03f15ac45e5f7dcf118d1e7ce1556339aa46/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala) to actually return a `Predicate` rather than a bare lambda function.  The code there already handles binding and evaluation.

**Comment by kiszk** [comment]
Thank you for pointing out `InterpretedPredicate`. Now, `newPredicate` always returns `Predicate` that can be executed by calling `eval()`.
It looks simpler and better than the first implementation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 2]
```diff
@@ -372,7 +374,7 @@ case class WholeStageCodegenExec(child: SparkPlan) extends UnaryExecNode with Co
     try {
       CodeGenerator.compile(cleanedSource)
     } catch {
-      case e: Exception if !Utils.isTesting && sqlContext.conf.wholeStageFallback =>
+      case e: JaninoRuntimeException if !Utils.isTesting && sqlContext.conf.wholeStageFallback =>
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: This change is useless. `CodeGenerator.doCompile` catches `Exception` and re-throw `Exception`.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: Can other exceptions be thrown during compiling?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 3]
```diff
@@ -951,10 +965,10 @@ object CodeGenerator extends Logging {
       evaluator.cook("generated.java", code.body)
       recordCompilationStats(evaluator)
     } catch {
-      case e: Exception =>
+      case e: JaninoRuntimeException =>
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Looks like `CompileException` can be thrown from janino?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 4]
```diff
@@ -355,7 +357,21 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
 
   protected def newPredicate(
       expression: Expression, inputSchema: Seq[Attribute]): GenPredicate = {
-    GeneratePredicate.generate(expression, inputSchema)
+    try {
+      GeneratePredicate.generate(expression, inputSchema)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Shall we only do this fallback if `sqlContext.conf.wholeStageFallback` is turned on?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 5]
```diff
@@ -32,6 +32,7 @@ import org.apache.spark.sql.types.LongType
 import org.apache.spark.util.ThreadUtils
 import org.apache.spark.util.random.{BernoulliCellSampler, PoissonSampler}
 
+
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: nit: Extra space line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 6]
```diff
@@ -355,7 +357,21 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
...
+      case e: JaninoRuntimeException =>
+        val str = expression.toString
+        val logMessage = if (str.length > 256) {
+          str.substring(0, 256 - 3) + "..."
+        } else {
+          str
+        }
+        logWarning(s"Codegen disabled for this expression:\n $logMessage")
+        InterpretedPredicate.create(expression, inputSchema)
+      case e: Exception =>
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: This case can be removed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 4 above)
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: It is good to control it using an option. This is not a part of the whole-stage codegen.
Is it better to use `sqlContext.conf.wholeStageFallback` or add `sqlContext.conf.codegenFallback`?
What do you think?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: I am wondering if it makes sense that `wholeStageFallback` is false and this new option is true, or vice verse.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 7]
```diff
@@ -20,20 +20,22 @@ package org.apache.spark.sql.catalyst.expressions
...
+  def create(expression: Expression, inputSchema: Seq[Attribute]): InterpretedPredicate =
     create(BindReferences.bindReference(expression, inputSchema))
 
-  def create(expression: Expression): (InternalRow => Boolean) = {
-    (r: InternalRow) => expression.eval(r).asInstanceOf[Boolean]
-  }
+  def create(expression: Expression): InterpretedPredicate = new InterpretedPredicate(expression)
 }
 
+class InterpretedPredicate(expression: Expression) extends GenPredicate {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Is this change necessary?

Btw, let `InterpretedPredicate` extends `GenPredicate` looks weird logically.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 5 above)
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Good catch, done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 3 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Thank you for pointing out . Now, `CompileException` can be caught.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 4 above)
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Sure, now look at `sqlContext.conf.wholeStageFallback`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 6 above)
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Yeah, done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala (hunk 2 above)
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: Sure, reverted

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 7 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: First, this change is necessary to return `Predicate` for `InterpretedPredicate`.

For `GenPredicate`, I used the same naming convention in `SparkPlan.scala`. If we use `Predicate`, we will have a conflict with others. What name would it be better instead of `GenPredicate`?

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 7 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: `BasePredicate`?

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Thanks, looks good. Done

**Comment by kiszk** [comment]
ping @marmbrus

**Comment by kiszk** [comment]
@marmbrus, could you please take a look at this PR for fixing a problem?

**Comment by kiszk** [comment]
ping @marmbrus, @davies

**Comment by kiszk** [comment]
ping @marmbrus, @davies

**Comment by kiszk** [comment]
ping @marmbrus

**Comment by kiszk** [comment]
@marmbrus could you please take a look?

**Comment by kiszk** [comment]
@marmbrus could you please take a look?

**Comment by kiszk** [comment]
@marmbrus could you please take a look?

**Comment by kiszk** [comment]
ping @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 8]
```diff
@@ -20,20 +20,22 @@ package org.apache.spark.sql.catalyst.expressions
...
     create(BindReferences.bindReference(expression, inputSchema))
 
-  def create(expression: Expression): (InternalRow => Boolean) = {
-    (r: InternalRow) => expression.eval(r).asInstanceOf[Boolean]
-  }
+  def create(expression: Expression): InterpretedPredicate = new InterpretedPredicate(expression)
 }
 
+class InterpretedPredicate(expression: Expression) extends BasePredicate {
+  def eval(r: InternalRow): Boolean = expression.eval(r).asInstanceOf[Boolean]
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: nit: **override** def eval...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 9]
```diff
@@ -20,20 +20,22 @@ package org.apache.spark.sql.catalyst.expressions
...
+  def create(expression: Expression, inputSchema: Seq[Attribute]): InterpretedPredicate =
     create(BindReferences.bindReference(expression, inputSchema))
 
-  def create(expression: Expression): (InternalRow => Boolean) = {
-    (r: InternalRow) => expression.eval(r).asInstanceOf[Boolean]
-  }
+  def create(expression: Expression): InterpretedPredicate = new InterpretedPredicate(expression)
 }
 
+class InterpretedPredicate(expression: Expression) extends BasePredicate {
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: nit: **case** class

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 10]
```diff
@@ -899,8 +902,20 @@ object CodeGenerator extends Logging {
...
    * Compile the Java source code into a Java class, using Janino.
    */
-  def compile(code: CodeAndComment): GeneratedClass = {
+  def compile(code: CodeAndComment): GeneratedClass = try {
     cache.get(code)
+  } catch {
+    // Cache.get() may wrap the original exception. See the following URL
+    // http://google.github.io/guava/releases/14.0/api/docs/com/google/common/cache/
+    //   Cache.html#get(K,%20java.util.concurrent.Callable)
+    case e : UncheckedExecutionException =>
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: You can use the following simple codes:
```
case e@(_: UncheckedExecutionException | _: ExecutionError) =>
    throw e.getCause
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 11]
```diff
@@ -353,9 +356,28 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
...
+    InterpretedPredicate.create(expression, inputSchema)
+  }
+
   protected def newPredicate(
       expression: Expression, inputSchema: Seq[Attribute]): GenPredicate = {
-    GeneratePredicate.generate(expression, inputSchema)
+    try {
+      GeneratePredicate.generate(expression, inputSchema)
+    } catch {
+      case e: JaninoRuntimeException if sqlContext == null || sqlContext.conf.wholeStageFallback =>
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: nit: It's better to add a debug log to print `e`. you can merge two cases like this:
```
      case e@(_: JaninoRuntimeException | _: CompileException)
        if sqlContext == null || sqlContext.conf.wholeStageFallback =>
        logDebug(e.getMessage, e)
        genInterpretedPredicate(expression, inputSchema)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 12]
```diff
@@ -353,9 +356,28 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
     GenerateMutableProjection.generate(expressions, inputSchema, useSubexprElimination)
   }
 
+  private def genInterpretedPredicate(
+      expression: Expression, inputSchema: Seq[Attribute]): InterpretedPredicate = {
+    val str = expression.toString
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: I think Expression.toString will truncate too big expression. Right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 13]
```diff
@@ -951,10 +966,14 @@ object CodeGenerator extends Logging {
...
-      case e: Exception =>
+      case e: JaninoRuntimeException =>
         val msg = s"failed to compile: $e\n$formatted"
         logError(msg, e)
-        throw new Exception(msg, e)
+        throw new JaninoRuntimeException(msg, e)
+      case e: CompileException =>
+        val msg = s"failed to compile: $e\n$formatted"
+        logError(msg, e)
+        throw new CompileException(msg, e.asInstanceOf[CompileException].getLocation)
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: `throw new CompileException(msg, e.getLocation, e)`

**Review by zsxwing — COMMENTED** [review]
I made one pass. This looks good. Most of my comments are style issues.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 12 above)
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Yeah, `Expression.toString` truncates something. However, it does not work for this case. Thus, I did not change here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 11 above)
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Yes, done. Thank you.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 8 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Sure, done for both.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 13 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Good catch, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 10 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I see, done. thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 14]
```diff
@@ -899,8 +902,16 @@ object CodeGenerator extends Logging {
...
+  def compile(code: CodeAndComment): GeneratedClass = try {
     cache.get(code)
+  } catch {
+    // Cache.get() may wrap the original exception. See the following URL
+    // http://google.github.io/guava/releases/14.0/api/docs/com/google/common/cache/
+    //   Cache.html#get(K,%20java.util.concurrent.Callable)
+    case e @ (_: UncheckedExecutionException | _: ExecutionError) =>
+      val excChains = ExceptionUtils.getThrowables(e)
+      val exc = if (excChains.length == 1) excChains(0) else excChains(excChains.length - 2)
+      throw exc
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Why not use `e.getCause`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 11 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Forgot this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 14 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Good catch, done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 11 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: @kiszk could you fix this as well?

Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: @zsxwing Sorry, I made mistake. Now, I pushed it.

**Comment by zsxwing** [comment]
LGTM. Thanks! Merging to master.

**Comment by dongjoon-hyun** [comment]
Hi, All.
I'm wondering if it is too late for Spark 2.2.0 to include this.
Is this too risky for that?

**Comment by poplav** [comment]
Hi, All.  I am trying to get this included into Spark 2.1.1.  I opened a PR https://github.com/apache/spark/pull/18942.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 15]
```diff
@@ -353,9 +356,27 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
...
+  }
+
   protected def newPredicate(
       expression: Expression, inputSchema: Seq[Attribute]): GenPredicate = {
-    GeneratePredicate.generate(expression, inputSchema)
+    try {
+      GeneratePredicate.generate(expression, inputSchema)
+    } catch {
+      case e @ (_: JaninoRuntimeException | _: CompileException)
+          if sqlContext == null || sqlContext.conf.wholeStageFallback =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: `sqlContext.conf.wholeStageFallback` is almost useless here, because almost all the cases this will be done in executors.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Pretty risky if we always fallback. This might hide bugs.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Let me fix it by another PR.
